muffnn/autoencoder/autoencoder.py

```diff
@@ -0,0 +1,686 @@
+            if self._categorical_indices is not None:
+                min_float32 = tf.constant(tf.float32.min)
+                for i, begin, size in zip(
+                        range(self._categorical_indices.shape[0]),
+                        self._categorical_indices[:, 0],
+                        self._categorical_indices[:, 1]):
+                    scores += tf.nn.softmax_cross_entropy_with_logits(
+                        tf.slice(t, [begin, 0], [size, -1]),
+                        tf.slice(tf.transpose(self._input_values),
+                                 [begin, 0], [size, -1]),
+                        dim=0)
+
+                    # This one is painful. TensorFlow does not, AFAIK,
+                    # support assignments to Tensors that come out of
+                    # operations (it does support assignments to Variables).
+                    # So I am using a precomputed mask to to update certain
+                    # values. Because softmax has a sumexp operation,
+                    # you have to set the elements not updated to logits
+                    # equal to the most negative float. Then they come out
+                    # to zero after the exp and so fall out of the sumexp.
+                    msk = tf.transpose(self._categorical_msks[i, :, :])
+                    softmax = tf.nn.softmax(
+                        msk * t + (1.0 - msk) * min_float32,
+                        dim=0)
+                    t = (1.0 - msk) * t + msk * softmax
+
+            # Discrete 0/1 stuff.
+            # Sigmoid output w/ cross-entropy.
+            if self._binary_indices is not None:
+                tsub = tf.gather(t, self._binary_indices)
+                isub = tf.gather(tf.transpose(self._input_values),
+                                 self._binary_indices)
+                scores += tf.reduce_sum(
+                    tf.nn.sigmoid_cross_entropy_with_logits(tsub, isub),
+                    reduction_indices=[0])
+
+                binary_msk = tf.transpose(self._binary_msk)
+                t = ((1.0 - binary_msk) * t +
+                     binary_msk * tf.nn.sigmoid(t))
+
+            # Anything left uses the default metric.
+            if self._default_indices is not None:
+                default_msk = tf.transpose(self._default_msk)
+
+                if self.metric == 'mse':
+                    if self.output_activation is not None:
+                        t = ((1.0 - default_msk) * t +
+                             default_msk * self.output_activation(t))
+                    tsub = tf.gather(t, self._default_indices)
+                    isub = tf.gather(tf.transpose(self._input_values),
+                                     self._default_indices)
+                    diff = isub - tsub
+                    scores += tf.reduce_sum(tf.square(diff),
+                                            reduction_indices=[0])
+                elif self.metric == 'cross-entropy':
```

Comment:
You could, but then you are doing a bunch of extra work on your data for no reason (multiplying by masks, potentially in efficient gather operations over arrays, etc.). I think the multiple cases make sense here. Also, if you look carefully, the code itself is not duplicated. The code that does the mixed metric case is pretty different since it works on transposed stuff and has `tf.gather` calls.